proposals/inline-classes.md

```diff
@@ -0,0 +1,296 @@
+- Inline enum classes 
+    - Int enum for [Android IntDef](https://developer.android.com/reference/android/support/annotation/IntDef.html)
+    - String enum for Kotlin/JS (see [WebIDL enums](https://www.w3.org/TR/WebIDL-1/#idl-enums))
+- Units of measurement
+- Result type (aka Try monad) [KT-18608](https://youtrack.jetbrains.com/issue/KT-18608)
+- Inline property delegates
+```kotlin
+class A {
+    var something by InlinedDelegate(Foo()) // no actual instantiation of `InlinedDelegate`
+}
+
+
+inline class InlinedDelegate<T>(var node: T) {
+    operator fun setValue(thisRef: A, property: KProperty<*>, value: T) {
+        if (node !== value) {
+            thisRef.notify(node, value)
+        }
+        node = value
+    }
+
+    operator fun getValue(thisRef: A, property: KProperty<*>): T {
+        return node
+    }
+}
+```
+
+- Inline wrappers
+    - Typed wrappers
+    ```kotlin
+    inline class Name(private val s: String)
+    inline class Password(private val s: String)
+    
+    fun foo() {
+        var n = Name("n") // no actual instantiation, on JVM type of `n` is String
+        val p = Password("p")
+        n = "other" // type mismatch error
+        n = p // type mismatch error
+    }
+    ```
+
+    - API refinement
+    ```
+    // Java
+    public class Foo {
+        public Object[] objects() { ... }
+    }
+    
+    // Kotlin
+    inline class RefinedFoo(val f: Foo) {
+        inline fun <T> array(): Array<T> = f.objects() as Array<T>
+    }
+    ``` 
+
+## Description
+
+Inline classes are declared using soft keyword `inline` and must have a single property:
+```kotlin
+inline class Foo(val i: Int)
+```
+Property `i` defines type of the underlying runtime representation for inline class `Foo`, while at compile time type will be `Foo`.
+
+From language point of view, inline classes can be considered as restricted classes, they can declare various members, operators, 
+have generics. 
+
+Example:
+```kotlin
+inline class Name(val s: String) : Comparable<Name> {
+    override fun compareTo(other: Name): Int = s.compareTo(other.s)
+    
+    fun greet() {
+        println("Hello, $s")
+    }
+}    
+
+fun greet() {
+    val name = Name("Kotlin") // there is no actual instantiation of class `Name`
+    name.greet() // method `greet` is called as a static method
+}
+```
+
+## Current limitations
+
+Currently, inline classes must satisfy the following requirements:
+
+- Inline class must have a public primary constructor with a single value parameter
+- Inline class must have a single read-only (`val`) property as an underlying value, which is defined in primary constructor
+- Underlying value cannot be of the same type that is containing inline class
+- Inline class with undefined (recursively defined) generics, e.g. generics with an upper bound equal to the class, is prohibited
+    ```kotlin
+    inline class A<T : A<T>>(val x: T) // error
+    ```
+- Inline class cannot have `init` block
+- Inline class must be final
```

Comment:
Why would open inline classes not be allowed?
Couldn't they be extended by other inline classes that would only be additional API wrappers that could add more restrictions and more functions?

Reply:
How runtime should resolve which subtype of inline class it deals with?

Reply:
@Miha-x64 The inline classes have a boxed (non inline) type for cases where you use generics or super types like `Any`, so either it can be done statically, without autoboxing, or it is boxed, so behaves like a regular class, until it reaches an auto-unboxing place again.

To make an analogy to something existing, it would work the same as primitive number types, who have their boxed types extending the `Number` class.

Reply:
@LouisCAD to assign Int value to a Number variable, you must box it, which eliminates the whole purpose of inline classes.

Reply:
@Miha-x64 Does boxing an `Int` somewhere in your program eliminates the whole purpose of primitive, non boxed integers in your whole program? I don't think so, and the same applies for inline classes.
If you _always_ use them in ways that force autoboxing, then, yes, its pointless, but you'll usually not write autoboxing code everywhere.